Modify label_tag_mapping_add to accommodate scoped entity type.

djberg96 opened the pull request:
This PR is based on the changes in https://github.com/ManageIQ/manageiq/pull/14288. In that PR the mappable types are scoped by provider, e.g. "Amazon::Vm" or "Kubernetes::ContainerNode".

Instead of automatically using the hard coded AUTOTAG_PREFIX, it parses the provider name from the selected item if possible. The AUTOTAG_PREFIX is still the default value, which preserves backwards compatibility.

cben commented on app/controllers/ops_controller/settings/label_tag_mapping.rb:
```diff
@@ -128,8 +128,21 @@ def lt_map_get_form_vars
+      prefix, entity = entity.split('::')
+      # Backwards compatibility
+      if prefix && entity
+        prefix.downcase!
```
Consider just having lowercase prefixes in MAPPABLE_ENTITIES.
Capitalized with `::` looks like provider class name, but it's not exactly, IMO more confusing then helpful.

But then you might need upcase to display as "Amazon/Vm" in UI?  BTW, will we want translation of Amazon.
BTW, your screenshot shows "Amazon/Vm" with slash, where does `::` -> `/` replacement happen?

cben commented on app/controllers/ops_controller/settings/label_tag_mapping.rb:
```diff
@@ -128,8 +128,21 @@ def lt_map_get_form_vars
+      else
+        prefix = ContainerLabelTagMapping::AUTOTAG_PREFIX
+      end
+    end
```
This method was already too long, please extract a method to compute category name (or some other split).

cben commented:
4 ems_container_controller_spec, container_node_controller_spec failures unrelated, fixed on ui-classic master.
2 label_tag_mapping_spec failures relevant of course.

djberg96 commented:
@cben, Yeah, the UI depends on the MAPPABLE_ENTITIES, so I didn't want to downcase those. I'm not sure where the "/" to "::" conversion happens right off. I remember seeing it somewhere buried in the code, but it's definitely what I saw in the logs.

I've refactored the method a bit and added some custom specs. I'm not sure what Rubocop is whining about exactly.

djberg96 commented:
@miq-bot remove_label wip

djberg96 commented:
@blomquisg Ok, ready to go.

blomquisg commented:
https://github.com/ManageIQ/manageiq/pull/14288 is merged now.

blomquisg commented:
@miq-bot assign dclarizio

blomquisg commented:
@miq-bot add_label enhancement, cloud/compute

blomquisg commented:
@miq-bot add_label compute/cloud

blomquisg commented:
@miq-bot remove_label "pending core"

blomquisg commented:
@miq-bot remove_label pending core

djberg96 commented:
@dclarizio @AparnaKarve  Look alright? Can we merge?

AparnaKarve commented:
@djberg96 Tested successfully in the UI.

Other than the PR #, LGTM :)

dclarizio commented:
@AparnaKarve I see "container" on a lot of those screens, can you comment on what else needs to be changed?  Thx, Dan

AparnaKarve commented:
> I see "container" on a lot of those screens, can you comment on what else needs to be changed? 

@dclarizio Good point.

@djberg96 So looks like this is not applicable to `Containers` alone anymore, the `entity` could also be `Amazon` now.

To keep it more generic we could just lose the `Containers` word in the haml.
More like -- 
`Choose an entity and label` in the first screenshot, and
`Entity` and `Label` in the table in the second screenshot

screenshot#1
<img width="848" alt="screen shot 2017-03-16 at 11 28 24 am" src="https://cloud.githubusercontent.com/assets/1538216/24012670/082dc5da-0a3c-11e7-93b9-a3395662d92c.png">

screenshot#2
<img width="860" alt="screen shot 2017-03-16 at 11 28 12 am" src="https://cloud.githubusercontent.com/assets/1538216/24012675/0ca52b9e-0a3c-11e7-9ec5-3d56f597ea67.png">

djberg96 commented:
@AparnaKarve @dclarizio Yes, those should be reworked but I would like to save it for a future PR. In the next iteration, the plan is to add another selection field (provider). Can that wait until then?
